### ISSUE TITLE
no-member when a function returning multiple types returns a value that has __getattr__

### ISSUE THREAD

**Issue by mattwala** [PyCQA/pylint#3388]
### Steps to reproduce

```python
class X:
    def __getattr__(self, attr):
        pass


def f(y):
    if y:
        return 9
    else:
        return X()


def g(y):
    f(y).attr
```

### Current behavior

```
$ pylint -E test.py
************* Module test
test.py:17:4: E1101: Instance of 'int' has no 'attr' member (no-member)
```

### Expected behavior

I'm not sure. However, if you replace the definition of X with
```python
class X:
    attr = None
```
then there's no error, so it seems like inconsistent behavior.

### pylint --version output

$ pylint --version
```
pylint 2.4.4
astroid 2.3.3
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 14:38:56) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
```

cc: @inducer

**Comment by kthy** [comment]
I think I'm seeing the same error here but in a different context. In the following code snippet, `self.modifiers` is a `Modifier` list.

```python
class Modifier(Enum):
    ABSTRACT = "abstract"
    PARTIAL = "partial"
    SEALED = "sealed"
    STATIC = "static"

…
def __repr__(self):
    mods = " " * bool(self.modifiers)
    mods += " ".join([m.value for m in self.modifiers])
```

The last line is flagged with `pylint(no-member)`. However, the following variant do not show the error:

```python
def __repr__(self):
    mods = " " * bool(self.modifiers)
    mods = " ".join([m.value for m in self.modifiers])
```

My environment:
```bash
$ pylint --version
pylint 2.4.4
astroid 2.3.3
Python 3.6.9 (default, Nov  7 2019, 10:44:02)
[GCC 8.3.0]
```

**Comment by PCManticore** [comment]
@mattwala `pylint` doesn't have control flow inference, so it analyzes both branches of the `f()` function. A branch will return a concrete object that is an integer, so `pylint` knows it does not have an `attr` attribute and emits the error you are seeing. For `None` we have a special case which can be disabled via `--ignore-none=no` option, which will force `pylint` to trigger this error for `None` as well.

@kthy Thanks for the example, but I can't reproduce it especially since the example is not full. I'm not sure from where `modifiers` is coming from. Please open a separate issue and maybe there's something we can do.

**Comment by kthy** [comment]
Thanks, I'll open an issue if I can reproduce with a MWE.

**Comment by mattwala** [comment]
@PCManticore Thanks for the reply. I am still curious why there isn't an error in the second case I mentioned, when you replace the definition of the class with one that doesn't have `__getattr__`.

**Comment by PCManticore** [comment]
@mattwala As already mentioned `For None we have a special case which can be disabled via --ignore-none=no option, which will force pylint to trigger this error for None as well.` When we encounter `None` as the owner of an attribute, by default we skip emitting `no-member`, unless disabled via `--ignore-none=no`. I don't know *why* we do that, this behaviour was enabled before  I started maintaining pylint: https://github.com/PyCQA/pylint/commit/e40704c23b63e3a63a9e46a40548a75d75ed48bc

**Comment by mattwala** [comment]
@PCManticore Thanks. However, in the situation I'm talking about, the function doesn't return `None`, it returns an object of type `X` with an attribute named `attr`. I'm not sure what the `None` check has to do with it.
